engines/router/missionctl/server/grpc/server.go

```diff
@@ -0,0 +1,157 @@
+package grpc
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/caraml-dev/turing/engines/router/missionctl"
+	"github.com/caraml-dev/turing/engines/router/missionctl/errors"
+	"github.com/caraml-dev/turing/engines/router/missionctl/instrumentation/metrics"
+	"github.com/caraml-dev/turing/engines/router/missionctl/instrumentation/tracing"
+	"github.com/caraml-dev/turing/engines/router/missionctl/log"
+	"github.com/caraml-dev/turing/engines/router/missionctl/log/resultlog"
+	"github.com/caraml-dev/turing/engines/router/missionctl/turingctx"
+	upiv1 "github.com/caraml-dev/universal-prediction-interface/gen/go/grpc/caraml/upi/v1"
+	"github.com/gojek/fiber"
+	fibergrpc "github.com/gojek/fiber/grpc"
+	"github.com/opentracing/opentracing-go"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/reflection"
+)
+
+const turingReqIDHeaderKey = "Turing-Req-ID"
+const tracingComponentID = "grpc_handler"
+
+type UPIServer struct {
+	upiv1.UnimplementedUniversalPredictionServiceServer
+
+	missionControl missionctl.MissionControlUPI
+	port           int
+}
+
+func NewUPIServer(mc missionctl.MissionControlUPI, port int) *UPIServer {
+	return &UPIServer{
+		missionControl: mc,
+		port:           port,
+	}
+}
+
+func (us *UPIServer) Run() {
+	s := grpc.NewServer()
+	upiv1.RegisterUniversalPredictionServiceServer(s, us)
+	reflection.Register(s)
+	listener, err := net.Listen("tcp", fmt.Sprintf(":%d", us.port))
+	if err != nil {
+		log.Glob().Errorf("Failed to listen on port %d: %s", us.port, err)
+		return
+	}
+
+	errChan := make(chan error, 1)
+	stopChan := make(chan os.Signal, 1)
+
+	// bind OS events to the signal channel
+	signal.Notify(stopChan, syscall.SIGTERM, syscall.SIGINT)
+
+	go func() {
+		if err := s.Serve(listener); err != nil {
+			errChan <- err
+		}
+	}()
+
+	defer func() {
+		s.GracefulStop()
+	}()
+
+	// block until either OS signal, or server fatal error
+	select {
+	case err := <-errChan:
+		log.Glob().Errorf("Failed to start Turing Mission Control API: %s", err)
+	case <-stopChan:
+		log.Glob().Info("Signal to stop server")
+	}
+
+}
+
+func (us *UPIServer) PredictValues(ctx context.Context, req *upiv1.PredictValuesRequest) (
+	*upiv1.PredictValuesResponse, error) {
+	var predictionErr *errors.TuringError
+	defer metrics.GetMeasureDurationFunc(predictionErr, tracingComponentID)()
+
+	// Create context from the request context
+	ctx = turingctx.NewTuringContext(ctx)
+	// Create context logger
+	ctxLogger := log.WithContext(ctx)
+	defer func() {
+		_ = ctxLogger.Sync()
```

Comment:
Why do we need to `Sync` when exiting `PredictValues` ? Wouldn't that slow it down?

Reply:
hmm im not 100% sure on this, the http handler are behaving the same too.

It is [written in zap comment](https://github.com/uber-go/zap/blob/master/logger.go#L274) that application should handle it before exiting and also [sync in their quickstart](https://github.com/uber-go/zap#quick-start). 

But im not sure when should it be synced exactly.

Reply:
> It is [written in zap comment](https://github.com/uber-go/zap/blob/master/logger.go#L274) that application should handle it before exiting and also [sync in their quickstart](https://github.com/uber-go/zap#quick-start).

It's correct, you should call `Sync` explicitly when exiting the program (i.e. shutting down), other than that calling `Sync` should be avoided unless necessary. Depending on the logger sink, `Sync` might cause performance degradation. E.g. if the sink is the file, that means every call to `Sync` will write to the file.